ner-search/ner_ranker.py

```diff
@@ -0,0 +1,95 @@
+from collections import namedtuple
+from enum import Enum
+from typing import Dict, List, Optional
+
+import spacy
+from jina.executors.rankers import Match2DocRanker
+
+NEREntity = namedtuple('NEREntity', 'value type')
+
+
+class EntityMatchEnum(Enum):
+    NONE = 0  # no match
+    TYPE = 1  # matches on same type but diff value
+    FULL = 2  # same type and same value
+    FULL_ALL = 3  # all of them match
+
+
+class NERSpacyRanker(Match2DocRanker):
+    # TODO docstring
+    def __init__(
+        self,
+        model='en_core_web_lg',
+        factor_type_match=1.1,
+        factor_full_match=1.4,
+        *args,
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+        self.model = model
+        self.factor_full_match = factor_full_match
+        self.factor_type_match = factor_type_match
+
+    def post_init(self):
+        """Init the Spacy model"""
+        super(NERSpacyRanker, self).post_init()
+        # TODO check if isntalled, if not downlaod
+        self.nlp = spacy.load(self.model)
+
+    def score(
+        self,
+        old_matches_scores: List[List[float]],
+        queries_metas: List[Dict],
+        matches_metas: List[List[Dict]],
+    ) -> List[List[float]]:
+        """Calculate the negative Levenshtein distance
+
+        :param old_matches_scores: Contains old scores in a list
+        :param queries_metas: Dictionary containing all the query meta information requested by the `query_required_keys` class_variable.
+        :param matches_metas: List containing all the matches meta information requested by the `match_required_keys` class_variable. Sorted in the same way as `old_match_scores`
+        :return: An iterable of the new scores
+        """
+        return [
+            self._recalc_score_matches(matches_metas[i], old_matches_scores[i], query_meta)
+            for i, query_meta in enumerate(queries_metas)
+        ]
+
+    def _recalc_score_matches(self, matches_metas, old_matches_scores, query_meta):
+        query_ents = [NEREntity(value=e.text, type=e.label_) for e in self.nlp(query_meta['text'].strip()).ents]
```

Comment:
should be a parameter to the `_recalc_score_matches` function and only calculated once.

Reply:
It is calculated once per query

Reply:
This is actually only true for `traversal_paths=['r']`. In the moment you choose `traversal_paths=['c']` it is not true anymore. Anyhow, this is a problem of the current Executor/Driver design, so just ignore my comment.